### ISSUE TITLE
Default expiration setting

### ISSUE THREAD

**Issue by rugk** [keepassxreboot/keepassxc#2612]
[TIP]:  # ( Provide a general summary of the issue in the title above ^^ )
[TIP]:  # ( Do not include screenshots of your actual database! )

Add a default expiration setting for new password entries.

## Expected Behavior
[NOTE]: # ( If you're describing a bug, tell us what should happen )
[NOTE]: # ( If you're suggesting a change/improvement, tell us how it should work )

Add a setting to enable a default "expire time", e.g. "two weeks" or "3 months" relative to the creation of the password entry, that is set by default when you create a new password.

## Current Behavior
[NOTE]: # ( If describing a bug, tell us what happens instead of the expected behavior )
[NOTE]: # ( If suggesting a change/improvement, explain the difference from the current behavior )

Ýou always need to specify such an expiration manually.

## Context
[NOTE]: # ( How has this issue affected you? What are you trying to accomplish? )
[NOTE]: # ( Providing context helps us come up with a solution that is most useful in the real world )

Inspired by https://github.com/keepassxreboot/keepassxc/issues/551#issuecomment-445272773, who seems to not wanted to create a new issue for that suggestion.

**Comment by droidmonkey** [comment]
This should be a per-group setting.

**Comment by renyhp** [comment]
This would be wonderful. The expiration time should also automatically update, according to this setting, whenever the password is changed.

**Comment by jNullj** [comment]
I would love to see that feature, it really grinds my gears to change that on every entry and when updating the password.
I could do that if you want help. Please assign me to this issue.

Let me review the required work:

- [ ] Add new setting under group settings -> group named "default password expiration" which inherits from parent folders by default similar to "default auto-type".
- [ ] The new setting should have a disabled state. enabled state (default from parent) and value from the presets we use today.
- [ ] On new entry or editing old entry if the new setting is enabled the expiration of the password should be set automatically only when a password is added/changed to the entry.
- [ ] Update docs
- [ ] Update tests

Here is a question, should the user edit an entry with existing password and remove it (empty password) or create a new entry without a password what change should take place? Should the expiration automatically get de-activated for that entry or should it use the default value?
I suggest automatically removing the expiration as there is no logic in keeping track when there is no password and this would also make the "expired" search cluttered. But some users might want to control this if they use the expiration for something other then the password field (maybe attached file with a secret that needs renewal?) What do you think?

Regarding default value to use in the root folder i suggest keeping this setting disabled by default, do you know of any reasons to set a specific preset value?

**Comment by droidmonkey** [comment]
I recommend NOT working on this at all. Password expiration/rotation is no longer recommended, and in the rare event you need to set one then you do it. The edge cases and ambiguity you raised highlights the amount of logic and exceptions you'll have to bake into this to make it useful. Most likely it will remain disabled. 

If you do choose to spend time on this.... you should add this setting to the GROUP settings with inheritance. Very similar to the autotype and search settings. Since this is not in the KDBX standard, you'll have to use CustomData.

**Comment by rugk** [comment]
I tend to agree here even though I have created this issue *checks notes…* four years ago🫣

Things change and even NIST does not recommend wild rotating your password anymore.

That said with a password manager it is quite easy and maybe for some passwords it may (still) be needed. So I agree in group settings that would be enough.

After all it should not be so hard to implement/not a big maintenance cost, so having it is likely not bad.

**Comment by Kiread-work** [comment]
Unfortunately, I can't answer that question that way for conservative employers who want their password changed every 3 weeks.

thx

**Comment by Mathsterk** [comment]
I feel like this would be handy, not simply for wildly rotating password, but also for clearing out old accounts that may not be used any more. My use case may be a weird one, but i usually end up registering accounts for something, use it for 2-3 weeks, and then completely forget about it, and now i suddenly have a ton of old accounts to clear out. It would be nice to do an incremental cleanup instead of having to do tons of entries.

**Comment by droidmonkey** [comment]
Wouldn't you just have a ton of accounts that are now expired for which you have no idea which you want to keep? Just select a preset when you register the account.
